### PR TITLE
Add CurrentUserService

### LINKED ISSUE #330: Create new Service for work with HttpContext
New service must take HttpContext   from Controller and return different  data (role, id ....) for others service .

### PR DESCRIPTION
**Pull request by gingerbread-monster**
Adding helper service - `CurrentUserService`, that allows us to access properties of the current authorized user.

When we need to access user's information, all we need to do is this:
```cs
// _currentUserService is a service injected via class constructor
long accountId = _currentUserService.AccountId;
string email = _currentUserService.Email;
UserRole role = _currentUserService.Role;
```

By using this utility service we can replace this
https://github.com/ita-social-projects/WhatBackend/blob/a3680634f4db7a7c443084bfd5a83ba563a85682/CharlieBackend.Business/Services/AttachmentService.cs#L63

or this
https://github.com/ita-social-projects/WhatBackend/blob/a38473007e5fe294059a73543f59383a1df57399/CharlieBackend.Business/Services/DashboardService.cs#L240

with getters of the class as shown above.

---
Changes:
* Implemented `CurrentUserService`
* Updated controllers & services to use `ICurrentUserService` to access user's data.
* Updated unit tests.

### CODE REVIEW

File: CharlieBackend.Business/Services/CurrentUserService.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            get
+            {
+                string accountIdString = _httpContextAccessor
+                    .HttpContext
+                    .User
+                    ?.Claims
+                    ?.SingleOrDefault(claim => claim.Type == "AccountId")
+                    ?.Value;
+
+                long accountId;
```
Comment (andy-dendy-repo) [inline CharlieBackend.Business/Services/CurrentUserService.cs]: If this service is gonna be scoped, don't you think that it's better to store properties' data in private members to decrease the number of data evaluation to one? Like that:

```
private long _accountId = -1;

public long AccountId
        {
            get
            {
                 if(_accountId == -1)
                 {
                      ...
                 }
                 else
                 {
                     return _accountId;
                 }
        }
```

Comment (gingerbread-monster) [reply CharlieBackend.Business/Services/CurrentUserService.cs]: @andy-dendy-repo 
Optimized getters logic according to your comment in https://github.com/ita-social-projects/WhatBackend/pull/349/commits/21f38d99eecd6eb1980a0d461612db134929e9ae

File: CharlieBackend.Api.UnitTes/DashboardServiceTest.cs [hunk 2]
```diff
@@ -84,7 +81,9 @@ public async Task GetStudentsClassbook()
 
             _unitOfWorkMock.Setup(x => x.DashboardRepository).Returns(dashboardRepositoryMock.Object);
 
-            var dashboardService = new DashboardService(_unitOfWorkMock.Object);
+            var dashboardService = new DashboardService(
+                unitOfWork: _unitOfWorkMock.Object,
+                currentUserService: null);
```
Comment (mahotskuy) [inline CharlieBackend.Api.UnitTes/DashboardServiceTest.cs]: Create a field for _currentUserService. Insert mock object instead.

File: CharlieBackend.Business/Services/CurrentUserService.cs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            }
+        }
+
+        public string Email
+        {
+            get
+            {
+                if (_email is null)
+                {
+                    _email = _httpContextAccessor
```
Comment (mahotskuy) [inline CharlieBackend.Business/Services/CurrentUserService.cs]: Consider to create a private method to return appropriate claim and substitute common logic.

**Review by mahotskuy — APPROVED** [review]

**Review by QuestroL — APPROVED** [review]

**Review by andy-dendy-repo — APPROVED** [review]

File: CharlieBackend.Business/Services/CurrentUserService.cs (hunk 3 above)
Comment (gingerbread-monster) [reply CharlieBackend.Business/Services/CurrentUserService.cs]: @mahotskuy 
Fixed. Moved common getters' common logic of obtaining claim value into a private method.
In this commit: https://github.com/ita-social-projects/WhatBackend/pull/349/commits/3bf40030b640ffccad4698bb3f5bd75f6616c5ed

File: CharlieBackend.Api.UnitTes/DashboardServiceTest.cs (hunk 2 above)
Comment (gingerbread-monster) [reply CharlieBackend.Api.UnitTes/DashboardServiceTest.cs]: @mahotskuy 
If I understood you correctly, when writing tests, we need to provide (and mock) all service dependencies regardless of whether they are really used by service or not.
Fixed in this commit: https://github.com/ita-social-projects/WhatBackend/pull/349/commits/a080bd14acd9be9b2680d5942420645c64e7f09d

**Review by Hlib-mats — APPROVED** [review]

**Review by Serhii-Dudar — APPROVED** [review]

**Pull request merged**